src/modules/consumer.js

```diff
@@ -27,22 +30,54 @@ class Consumer {
    * @return {any}       object, string, number... the current received message
    */
   async checkRpc(messageProperties, queue, reply) {
-    if (messageProperties.replyTo) {
-      const options = {
-        correlationId: messageProperties.correlationId,
-        persistent: true,
-        durable: true,
-      };
-      this._connection.config.logger.debug({
-        message: `${loggerAlias} [${queue}][${messageProperties.replyTo}] > ${reply}`,
-        params: { content: reply },
-      });
+    if (!messageProperties.replyTo) {
+      return messageProperties;
+    }
+
+    const options = {
+      correlationId: messageProperties.correlationId,
+      persistent: true,
+      durable: true,
+    };
+    const serializedReply = parsers.out(reply, options);
+    await this.hooks.trigger(this, ConsumerHooks.beforeRpcReplyEvent, {
+      receiveProperties: messageProperties,
+      replyProperties: options,
+      queue,
+      reply,
+      serializedReply,
+      error: reply && reply.error ? reply.error : undefined,
+    });
+    logger.debug({
+      message: `${loggerAlias} [${queue}][${messageProperties.replyTo}] > ${reply}`,
+      params: { content: reply },
+    });
 
+    try {
       const defaultChannel = await this._connection.getDefaultChannel();
-      return await defaultChannel.sendToQueue(messageProperties.replyTo, parsers.out(reply, options), options);
-    }
+      const written = defaultChannel.sendToQueue(messageProperties.replyTo, serializedReply, options);
+      await this.hooks.trigger(this, ConsumerHooks.afterRpcReplyEvent, {
+        receiveProperties: messageProperties,
+        queue,
+        reply,
+        serializedReply,
+        replyProperties: options,
+        written,
+      });
 
-    return messageProperties;
+      return written;
+    } catch (error) {
+      await this.hooks.trigger(this, ConsumerHooks.afterRpcReplyEvent, {
```

Comment:
why don't you use a `finally` block instead of duplicating the code?

Reply:
It has an `extra` error parameter. I extracted it outside the block so it can be set in the catch and call `trigger` only once in the catch.

Did it similarly on other places where it matters.

Commit https://github.com/bringg/node-arnavmq/pull/67/commits/78716b7e8bae977e858474110210ce74fb55a2bf